op-node/rollup/interop/managed/system.go

```diff
@@ -378,60 +378,121 @@ func (m *ManagedMode) Reset(ctx context.Context, lUnsafe, xUnsafe, lSafe, xSafe,
+
+			logger.Info("Last valid L2 block ref", "valid", valid)
+			return valid, nil
+		}
+	} else if x < 0 {
+		logger.Warn("Latest unsafe block is older than target, using latest unsafe for search")
+		target = latestUnsafe.Number
 	}
 
-	// Start from latestUnsafe and go down to find the latest valid block, since L1 reorgs aren't usually that deep.
-	for n := latestUnsafe.Number; ; n-- {
-		if n == l2UnsafeTarget.Number-1 {
+	// In the following walkback loop, the following two cases are covered:
+	// 1. x == 0 or x < 1 (i.e. target == latestUnsafe), or
+	// 2. binary search returned offset == 0 (i.e. all blocks we checked are invalid), so we have to go from `target` backwards indefinitely
+	//    until we find a valid block
+	for n := target; ; n-- {
+		if n == target-1 {
 			logger.Warn("No valid unsafe block found up to target, searching further")
 		}
-		// Exit loop if context is done - other errors are only logged and loop continues.
-		if cerr := ctx.Err(); cerr != nil {
-			return eth.L2BlockRef{}, fmt.Errorf("context done: %w", cerr)
-		}
 
-		current, err := m.l2.L2BlockRefByNumber(ctx, n)
+		isInvalid, valid, err := m.isInvalidBlock(ctx, logger, n)
 		if err != nil {
-			logger.Warn("Failed to get L2 block ref, trying previous", "err", err, "current", n)
-			continue
+			return eth.L2BlockRef{}, err
 		}
 
-		logger := logger.New("current", current, "currentL1Origin", current.L1Origin)
-		// make sure L1 origin hasn't been reorged
-		l1Origin, err := m.l1.L1BlockRefByNumber(ctx, current.L1Origin.Number)
+		if !isInvalid {
+			logger.Info("Fould last valid block", "valid", valid)
+			return valid, nil
+		}
+	}
+}
+
+// localUnsafeSearch is a copy-paste from sort.Search, but also stops on and returns error
+func localUnsafeSearch(n int, f func(int) (bool, eth.L2BlockRef, error)) (int, error) {
+	// Define f(-1) == false and f(n) == true.
+	// Invariant: f(i-1) == false, f(j) == true.
+	i, j := 0, n
+	for i < j {
+		h := int(uint(i+j) >> 1) // avoid overflow when computing h
+		// i ≤ h < j
+		ok, _, err := f(h)
 		if err != nil {
-			logger.Warn("Failed to get origin L1 block ref, trying previous", "err", err)
-			continue
+			return -1, err
 		}
-		if l1Origin.Hash != current.L1Origin.Hash {
-			logger.Warn("L1 block was reorged on this block, trying previous", "newL1Origin", l1Origin)
-			continue
+		if !ok {
+			i = h + 1 // preserves f(i-1) == false
+		} else {
+			j = h // preserves f(j) == true
 		}
-		logger.Info("Latest valid L2 block found", "valid", current)
-		return current, nil
 	}
+	// i == j, f(i-1) == false, and f(j) (= f(i)) == true  =>  answer is i.
+	return i, nil
+}
+
+// isInvalidBlock returns:
+// - false: if the L1Origin field for the block belongs to the canonical chain; in this case we also return the block ref
+// - true: if the L1Origin field for the block is invalid/outdated/reorged
+func (m *ManagedMode) isInvalidBlock(ctx context.Context, logger log.Logger, blockNum uint64) (bool, eth.L2BlockRef, error) {
```

Comment:
I would find it cleaner to change this function to
```suggestion
func (m *ManagedMode) verifyBlock(ctx context.Context, logger log.Logger, blockNum uint64) (eth.L2BlockRef, error) {
```
so it's less specific to the one use case inside the binary search and then to `binarySearch` you can just pass
```go
func(i int) (bool, eth.L2BlockRef, error) {
    block, err := m.isInvalidBlock(ctx, logger, target+1+uint64(i))
    return block == (eth.L2BlockRef{}), block, err
}
```

Reply:
I changed this, but I personally prefer the initial version -- here is the diff https://github.com/ethereum-optimism/optimism/pull/16276/commits/abc3fa703e527a72d050edf84457f66d7fc54471

In any case both are fine.